sklearn/metrics/_regression.py

```diff
@@ -1188,11 +1246,12 @@ def r2_score(
         y_true, y_pred, sample_weight, multioutput
     )
 
-    dtype = _find_matching_floating_dtype(y_true, y_pred, sample_weight, xp=xp)
-
-    _, y_true, y_pred, multioutput = _check_reg_targets(
-        y_true, y_pred, multioutput, dtype=dtype, xp=xp
+    _, y_true, y_pred, sample_weight, multioutput, dtype = (
```

Comment:
```suggestion
    _, y_true, y_pred, sample_weight, multioutput, _ = (
```

the call to `sample_weight = column_or_1d(sample_weight, dtype=dtype)` can be updated to avoid passing the `dtype` parameter has the dtype handling has already happened at this point.

Reply:
> [...] the dtype handling has already happened at this point.

Is this because the `dtype` handling for `r2_score` is managed by the `_assemble_r2_explained_variance` function? If so, should we update:

```python
def _assemble_r2_explained_variance(
    numerator, denominator, n_outputs, multioutput, force_finite, xp, device
):
    """Common part used by explained variance score and :math:`R^2` score."""
    dtype = numerator.dtype
```
to use our own `_find_matching_floating_dtype` instead?